compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -1148,6 +1149,37 @@ pub fn typeid_for_instance<'tcx>(
                 tcx.mk_args_trait(invoke_ty, trait_ref.args.into_iter().skip(1));
             instance.args = instance.args.rebase_onto(tcx, impl_id, abstract_trait_args);
         }
+    } else if tcx.is_closure_like(instance.def_id()) {
+        // We're either a closure or a coroutine. Our goal is to find the trait we're defined on,
+        // instantiate it, and take the type of its only method as our own.
+        let closure_ty = instance.ty(tcx, ty::ParamEnv::reveal_all());
```

Comment:
Get the type from `instance.args.type_at(0)`, since this will not be correct for async closures (which aren't really supported yet, but I don't want the situation to get worse).

Reply:
I'll do that, but for my own understanding, what does `instance.ty` return for an async closure that isn't the type of the async closure? (and why?).

Reply:
Experimentally, `instance.args.type_at(0)` returns a type of `()` for real closures, so I think the recommendation to use `type_at(0)` may have been meant for a different location?

Reply:
> I'll do that, but for my own understanding, what does instance.ty return for an async closure that isn't the type of the async closure?

Sorry, it's not async closures per se. It's the coroutines that come from calling an async closures. Depending on if you call them for `AsyncFnOnce` or `AsyncFnMut`, you may get two coroutine types that will only differ by their [`kind_ty`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/sty/struct.CoroutineArgs.html#method.kind_ty). Unfortunately, the implementation of `Instance::ty` kinda totally yeets that `kind_ty` and replaces it. It's why we have hacks [like this](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_ty_utils/abi.rs.html#115), but we can side-step needing to do anything fancy here by just using `instance.args` which are correct.